File: common/src/main/scripts/lib/micropython/oc2/bus.py
import os
import time

from oc2 import blob as oc2_blob
from oc2 import channel as oc2_channel
from oc2 import ports as oc2_ports
from oc2 import socket as oc2_socket
from oc2.blob import Blob, PayloadChannel
from oc2.channel import Channel
from oc2.events import Events

SOCKET_PATH = "/run/oc2/bus"
PROTOCOL_VERSION = 1
TOKEN_LENGTH = 32
HELLO_TIMEOUT_MS = 5000
REQUEST_TIMEOUT_MS = 35000
MAX_EVENTS_PER_PUMP = 32
CONNECT_ATTEMPTS = 5
CONNECT_RETRY_MS = 200
MAX_REPLY_SIZE = oc2_channel.DEFAULT_MAX_FRAME

RPC_PORT_NAME = "oc2.rpc.0"
BLOB_PORT_NAME = "oc2.blob.0"
EVENT_PORT_NAME = "oc2.event.0"

ROLE_CODE = {"rpc": b"R", "blob": b"B", "event": b"E"}
DEFAULT_ROLES = ("rpc", "blob", "event")


# Device

def _parameter_name(parameter, index):
    return parameter.get("name") or ("arg" + str(index))


class Device:
    def __init__(self, bus, device):
        self.bus = bus
        self.device_id = device["deviceId"]
        self.type_names = list(device.get("typeNames") or [])
        self._methods = None

    def _method_list(self):
        if self._methods is None:
            self._methods = self.bus.methods(self.device_id)
        return self._methods

    def invoke(self, method_name, *args):
        return self.bus.invoke(self.device_id, method_name, *args)

    def __getattr__(self, item):
        if item.startswith("_"):
            raise AttributeError(item)
        for method in self._method_list():
            if method.get("name") == item:
                return lambda *args: self.bus.invoke(self.device_id, item, *args)
        raise AttributeError("device %s has no method %s" % (self.device_id, item))

    def __str__(self):
        out = []
        for method in self._method_list():
            out.append(method["name"])
            out.append("(")
            parameters = method.get("parameters") or []
            for i, p in enumerate(parameters):
                if i > 0:
                    out.append(", ")
                out.append(_parameter_name(p, i + 1))
                if p.get("type"):
                    out.append(": ")
                    out.append(p["type"])
            out.append(")")
            if method.get("returnType"):
                out.append(": ")
                out.append(method["returnType"])
            out.append("\n")

            if method.get("description"):
                out.append(method["description"])
                out.append("\n")

            for i, p in enumerate(parameters):
                if p.get("description"):
                    out.append("  ")
                    out.append(_parameter_name(p, i + 1))
                    out.append("  ")
                    out.append(p["description"])
                    out.append("\n")
        return "".join(out)


# Bus

def _copy_device(device):
    copy = {}
    for key, value in device.items():
        copy[key] = list(value) if isinstance(value, list) else value
    return copy


def _copy_devices(devices):
    return [_copy_device(device) for device in devices]


class DeviceBus:
    null = None

    def __init__(self, transport, rpc, payload, events):
        self.transport = transport
        self.rpc = rpc
        self.payload = payload
        self.events = events
        self.generation = None
        self.generation_confirmed = False
        self.device_list = None

    def close(self):
        if self.rpc:
            self.rpc.close()
        if self.payload:
            self.payload.close()
        if self.events:
            self.events.close()

    def flush(self):
        if self.transport != "ports":
            return
        self.rpc.reset()
        self.payload.reset()

    def _note_generation(self, envelope):
        gen = envelope.get("gen")
        if gen is None:
            if envelope.get("type") == "error":
                return
            raise Exception("host reply carried no bus generation")
        if gen != self.generation:
            self.generation = gen
            self.device_list = None
        self.generation_confirmed = True

    def _apply_event(self, event):
        if event.get("type") == "eventsDropped":
            self.device_list = None
            self.generation_confirmed = False
            return True

        if event.get("type") == "devicesChanged":
            if event.get("gen") != self.generation:
                self.generation = event.get("gen")
                self.device_list = None
            self.generation_confirmed = True
            return True
        return False

    def pump_events(self):
        if self.events is None:
            return 0

        count = 0
        for _ in range(MAX_EVENTS_PER_PUMP):
            event = self.events.poll()
            if event is None:
                break
            if self._apply_event(event):
                count += 1
        return count

    def wait_event(self, timeout=None, event_type=None):
        if self.events is None:
            raise Exception("this session has no event channel")

        while True:
            event = self.events.wait(timeout)
            if event is None:
                return None
            self._apply_event(event)
            if event_type is None or event.get("type") == event_type:
                return event

    def _request(self, message, expected):
        self.rpc.write(message)
        reply = self.rpc.read(REQUEST_TIMEOUT_MS)
        if reply is None:
            self.rpc.reset()
            if self.payload is not None:
                self.payload.reset()
            raise Exception("no reply from the host: timeout")
        self._note_generation(reply)
        if reply.get("type") == expected:
            return reply
        if reply.get("type") == "error":
            raise Exception(reply.get("data")
                            or "the host reported an error with no detail")
        raise Exception("unexpected message type: %s" % reply.get("type"))

    def _raw_list(self):
        self.pump_events()

        if self.device_list is not None and self.generation_confirmed:
            self.generation_confirmed = False
            return self.device_list

        self.flush()
        self.device_list = self._request({"type": "list"}, "list").get("data")
        return self.device_list

    def list(self):
        return _copy_devices(self._raw_list())

    def _lookup(self, matches):
        for _ in range(2):
            for device in self._raw_list():
                if matches(device):
                    return Device(self, _copy_device(device))
            if self.device_list is None:
                break  # was not cached, so looking again would return the same thing
            self.device_list = None
        return None

    def get(self, device_id):
        return self._lookup(lambda candidate: candidate.get("deviceId") == device_id)

    def find(self, type_name):
        return self._lookup(
            lambda candidate: type_name in (candidate.get("typeNames") or []))

    def methods(self, device_id):
        self.flush()
        return self._request({"type": "methods", "data": device_id}, "methods").get("data")

    def blob(self, data):
        return Blob(data)

    def invoke(self, device_id, method_name, *args):
        self.flush()

        parameters, payload = oc2_blob.extract(args)
        message = {"type": "invoke", "data": {
            "deviceId": device_id,
            "name": method_name,
            "parameters": parameters
        }}

        if payload is not None:
            if self.payload is None:
                raise Exception("this session has no payload channel; "
                                "open one with the blob role")
            self.payload.write(payload)
            message["blob"] = {"length": len(payload),
                               "checksum": oc2_blob.checksum(payload)}

        reply = self._request(message, "result")
        if reply.get("blob") is not None and self.payload is None:
            raise Exception("the host sent a payload, but this session has no payload channel")
        return oc2_blob.resolve(self.payload, reply)


# Transports

def _connect_ports():
    rpc = payload = events = None
    try:
        paths = {}
        for name in (RPC_PORT_NAME, BLOB_PORT_NAME, EVENT_PORT_NAME):
            paths[name] = oc2_ports.find(name)
            if paths[name] is None:
                raise Exception("no virtio port named %s was found" % name)

        rpc = Channel.open(paths[RPC_PORT_NAME])
        payload = PayloadChannel.open(paths[BLOB_PORT_NAME])
        events = Events.open(paths[EVENT_PORT_NAME])
    except Exception:
        for handle in (rpc, payload, events):
            if handle is not None:
                handle.close()
        raise

    return DeviceBus("ports", rpc, payload, events)


def _connect_socket(path, roles):
    for role in roles:
        if role not in ROLE_CODE:
            raise Exception("unknown role %s" % role)

    opened = []
    try:
        rpc_socket = oc2_socket.connect(path)
        opened.append(rpc_socket)

        rpc = Channel(rpc_socket, MAX_REPLY_SIZE)
        rpc.write_raw(ROLE_CODE["rpc"])
        rpc.write({"type": "hello", "version": PROTOCOL_VERSION, "roles": list(roles)})

        hello = rpc.read(HELLO_TIMEOUT_MS)
        if hello is None:
            raise Exception("the bus daemon did not answer")
        if hello.get("type") != "hello":
            raise Exception("the bus daemon refused the session: %s"
                            % (hello.get("data") or hello.get("type")))
        token = hello.get("token")
        if not isinstance(token, str) or len(token) != TOKEN_LENGTH:
            raise Exception("the bus daemon sent an unusable session token")

        token = token.encode()
        payload = events = None
        for role in roles:
            if role == "rpc":
                continue
            sock = oc2_socket.connect(path)
            opened.append(sock)
            attach = ROLE_CODE[role] + token
            if role == "blob":
                payload = PayloadChannel(sock)
                payload.write(attach)
            else:
                events = Events.from_stream(sock, MAX_REPLY_SIZE)
                events.channel.write_raw(attach)
    except Exception:
        for sock in opened:
            try:
                sock.close()
            except Exception:
                pass
        raise

    bus = DeviceBus("socket", rpc, payload, events)
    bus.generation = hello.get("gen")
    return bus


def connect(socket_path=None, roles=DEFAULT_ROLES):
    path = socket_path or os.getenv("OC2_BUS_SOCKET") or SOCKET_PATH
    use_socket = path not in ("none", "")

    socket_error = None
    direct_error = None

    for attempt in range(CONNECT_ATTEMPTS):
        if use_socket:
            try:
                return _connect_socket(path, roles)
            except Exception as e:
                socket_error = e

        try:
            return _connect_ports()
        except Exception as e:
            direct_error = e

        if attempt < CONNECT_ATTEMPTS - 1:
            time.sleep(CONNECT_RETRY_MS / 1000)

    if socket_error is not None:
        raise Exception("could not reach the bus daemon (%s), nor the ports directly (%s)"
                        % (socket_error, direct_error))
    raise direct_error

File: common/src/main/scripts/lib/lua/oc2/bus.lua
local Channel = require("oc2.channel")
local blob = require("oc2.blob")
local Events = require("oc2.events")
local ports = require("oc2.ports")
local socket = require("oc2.socket")
local time = require("posix.time")

local bus = {}

bus.socketPath = "/run/oc2/bus"
bus.protocolVersion = 1
bus.tokenLength = 32
bus.helloTimeout = 5000
bus.requestTimeout = 35000
bus.maxEventsPerPump = 32
bus.connectAttempts = 5
bus.connectRetryDelay = 200
bus.maxReplySize = Channel.defaultMaxFrame

local rpcPortName = "oc2.rpc.0"
local blobPortName = "oc2.blob.0"
local eventPortName = "oc2.event.0"

bus.rpcPortName = rpcPortName
bus.blobPortName = blobPortName
bus.eventPortName = eventPortName

local roleCode = { rpc = "R", blob = "B", event = "E" }
local defaultRoles = { "rpc", "blob", "event" }

bus.roleCode = roleCode
bus.defaultRoles = defaultRoles

-- Device

local function parameterName(p, index)
  return p.name or ("arg" .. index)
end

local Device = {}
local invokers = {}
local owners = setmetatable({}, {__mode = "k"})

local function busOf(device)
  return owners[device].bus
end

local function methodsOf(device)
  local owner = owners[device]
  if not owner.methods then
    owner.methods = busOf(device):methods(device.deviceId)
  end
  return owner.methods
end

function Device.new(bus_, device)
  owners[device] = { bus = bus_ }
  return setmetatable(device, Device)
end

function Device:invoke(methodName, ...)
  return busOf(self):invoke(self.deviceId, methodName, ...)
end

Device.__index = function(self, key)
  local direct = rawget(Device, key)
  if direct then
    return direct
  end

  local found = false
  for _, method in ipairs(methodsOf(self)) do
    if method.name == key then
      found = true
      break
    end
  end
  if not found then
    return nil
  end

  local invoker = invokers[key]
  if not invoker then
    invoker = function(device, ...)
      return Device.invoke(device, key, ...)
    end
    invokers[key] = invoker
  end
  return invoker
end

Device.__tostring = function(self)
  local out = {}
  local function put(...)
    for _, part in ipairs({...}) do
      out[#out + 1] = part
    end
  end

  for _, method in ipairs(methodsOf(self)) do
    put(method.name, "(")
    if method.parameters then
      for i, p in ipairs(method.parameters) do
        if i > 1 then
          put(", ")
        end
        put(parameterName(p, i))
        if p.type then
          put(": ", p.type)
        end
      end
    end
    put(")")
    if method.returnType then
      put(": ", method.returnType)
    end
    put("\n")

    if method.description then
      put(method.description, "\n")
    end

    if method.parameters then
      for i, p in ipairs(method.parameters) do
        if p.description then
          put("  ", parameterName(p, i), "  ", p.description, "\n")
        end
      end
    end
  end

  return table.concat(out)
end

-- Bus

local function copyDevice(device)
  local copy = {}
  for key, value in pairs(device) do
    if type(value) == "table" then
      local inner = {}
      for innerKey, innerValue in pairs(value) do
        inner[innerKey] = innerValue
      end
      copy[key] = inner
    else
      copy[key] = value
    end
  end
  return copy
end

local function copyDevices(devices)
  local result = {}
  for i, device in ipairs(devices) do
    result[i] = copyDevice(device)
  end
  return result
end

local DeviceBus = {}
DeviceBus.__index = DeviceBus

DeviceBus.null = require("cjson").null

function DeviceBus.new(transport, rpc, payload, events)
  return setmetatable({
    transport = transport, rpc = rpc, payload = payload, events = events,
  }, DeviceBus)
end

function DeviceBus:close()
  self.rpc:close()
  if self.payload then
    self.payload:close()
  end
  if self.events then
    self.events:close()
  end
end

function DeviceBus:flush()
  if self.transport ~= "ports" then
    return
  end
  self.rpc:reset()
  self.payload:reset()
end

local function parseError(result, reason)
  if not result then
    return "unexpected error: " .. tostring(reason or "unknown error")
  elseif result.type == "error" then
    return tostring(result.data or "the host reported an error with no detail")
  else
    return "unexpected message type: " .. tostring(result.type)
  end
end

local function noteGeneration(bus_, result)
  if result.type == "error" and result.gen == nil then
    return -- Daemon error
  end

  local gen = assert(result.gen, "host reply carried no bus generation")
  if gen ~= bus_.generation then
    bus_.generation = gen
    bus_.deviceList = nil
  end
  bus_.generationConfirmed = true
end

local function applyEvent(bus_, event)
  if event.type == "eventsDropped" then
    bus_.deviceList = nil
    bus_.generationConfirmed = false
    return true
  end

  if event.type == "devicesChanged" then
    if event.gen ~= bus_.generation then
      bus_.generation = event.gen
      bus_.deviceList = nil
    end
    bus_.generationConfirmed = true
    return true
  end

  return false
end

function DeviceBus:pumpEvents()
  if not self.events then
    return 0
  end

  local count = 0
  for _ = 1, bus.maxEventsPerPump do
    local event = self.events:poll()
    if not event then
      break -- nothing pending, or a frame we could not parse; either way, stop
    end
    if applyEvent(self, event) then
      count = count + 1
    end
  end
  return count
end

function DeviceBus:waitEvent(timeout, eventType)
  if not self.events then
    return nil, "this session has no event channel"
  end

  while true do
    local event, reason = self.events:wait(timeout)
    if not event then
      return nil, reason
    end

    applyEvent(self, event)
    if not eventType or event.type == eventType then
      return event
    end
  end
end

local function request(bus_, message, expected)
  bus_.rpc:write(message)
  local result, reason = bus_.rpc:read(bus.requestTimeout)
  if not result then
    bus_.rpc:reset()
    if bus_.payload then
      bus_.payload:reset()
    end
    return error("no reply from the host: " .. tostring(reason or "unknown"), 0)
  end
  noteGeneration(bus_, result)
  if result.type == expected then
    return result
  end
  return error(parseError(result, reason), 0)
end

local function cachesLocally(bus_)
  return bus_.transport == "ports"
end

local function rawList(bus_)
  bus_:pumpEvents()

  if cachesLocally(bus_) and bus_.deviceList and bus_.generationConfirmed then
    bus_.generationConfirmed = false
    return bus_.deviceList
  end

  bus_:flush()
  local result = request(bus_, { type = "list" }, "list")
  if cachesLocally(bus_) then
    bus_.deviceList = result.data
  end
  return result.data
end

function DeviceBus:list()
  return copyDevices(rawList(self))
end

local function lookup(bus_, matches)
  for _ = 1, 2 do
    for _, device in ipairs(rawList(bus_)) do
      if matches(device) then
        return Device.new(bus_, copyDevice(device))
      end
    end

    if not bus_.deviceList then
      break -- was not cached, so looking again would return the same thing
    end
    bus_.deviceList = nil
  end
end

function DeviceBus:get(deviceId)
  local device = lookup(self, function(candidate)
    return candidate.deviceId == deviceId
  end)
  if device then
    return device
  end

  return nil, "no device with id [" .. deviceId .. "]"
end

function DeviceBus:find(deviceTypeName)
  local device = lookup(self, function(candidate)
    if candidate.typeNames then
      for _, typeName in ipairs(candidate.typeNames) do
        if typeName == deviceTypeName then
          return true
        end
      end
    end
    return false
  end)
  if device then
    return device
  end

  return nil, "no device of type [" .. deviceTypeName .. "]"
end

function DeviceBus:methods(deviceId)
  self:flush()
  return request(self, { type = "methods", data = deviceId }, "methods").data
end

function DeviceBus:blob(data)
  return blob.wrap(data)
end

function DeviceBus:invoke(deviceId, methodName, ...)
  self:flush()

  local parameters, payload = blob.extract(...)
  local message = { type = "invoke", data = {
    deviceId = deviceId,
    name = methodName,
    parameters = parameters
  }}

  if payload then
    if not self.payload then
      error("this session has no payload channel; open one with the blob role", 0)
    end
    self.payload:write(payload)
    message.blob = { length = #payload, checksum = blob.checksum(payload) }
  end

  local result = request(self, message, "result")
  if result.blob and not self.payload then
    error("the host sent a payload, but this session has no payload channel", 0)
  end
  return blob.resolve(self.payload, result)
end

-- Transports

local function connectPorts()
  local rpcPath = ports.find(rpcPortName)
  if not rpcPath then
    return nil, "no virtio port named " .. rpcPortName .. " was found"
  end
  local blobPath = ports.find(blobPortName)
  if not blobPath then
    return nil, "no virtio port named " .. blobPortName .. " was found"
  end
  local eventPath = ports.find(eventPortName)
  if not eventPath then
    return nil, "no virtio port named " .. eventPortName .. " was found"
  end

  local rpc, rpcReason = Channel.open(rpcPath)
  if not rpc then
    return nil, tostring(rpcReason)
  end

  local payload, payloadReason = blob.open(blobPath)
  if not payload then
    rpc:close()
    return nil, tostring(payloadReason)
  end

  local events, eventReason = Events.open(eventPath)
  if not events then
    rpc:close()
    payload:close()
    return nil, tostring(eventReason)
  end

  return DeviceBus.new("ports", rpc, payload, events)
end

local function connectSocket(path, roles)
  for _, role in ipairs(roles) do
    if not roleCode[role] then
      return nil, "unknown role " .. tostring(role)
    end
  end

  local opened = {}
  local function abandon(reason)
    for _, fd in ipairs(opened) do
      socket.close(fd)
    end
    return nil, reason
  end

  local function dial()
    local fd, reason = socket.connect(path)
    if not fd then
      return nil, reason
    end
    opened[#opened + 1] = fd
    return fd
  end

  local rpcFd, rpcReason = dial()
  if not rpcFd then
    return abandon(tostring(rpcReason))
  end

  local rpc = Channel.fromFd(rpcFd, bus.maxReplySize)
  local sent, sendReason = pcall(function()
    rpc:writeRaw(roleCode.rpc)
    rpc:write({ type = "hello", version = bus.protocolVersion, roles = roles })
  end)
  if not sent then
    return abandon(tostring(sendReason))
  end

  local hello, helloReason = rpc:read(bus.helloTimeout)
  if not hello then
    return abandon("the bus daemon did not answer: " .. tostring(helloReason))
  end
  if hello.type ~= "hello" then
    return abandon("the bus daemon refused the session: " .. parseError(hello))
  end
  if type(hello.token) ~= "string" or #hello.token ~= bus.tokenLength then
    return abandon("the bus daemon sent an unusable session token")
  end

  local payload, events
  for _, role in ipairs(roles) do
    if role ~= "rpc" then
      local fd, reason = dial()
      if not fd then
        return abandon(tostring(reason))
      end
      local attached, attachReason = pcall(function()
        Channel.fromFd(fd):writeRaw(roleCode[role] .. hello.token)
      end)
      if not attached then
        return abandon(tostring(attachReason))
      end
      if role == "blob" then
        payload = blob.fromFd(fd)
      else
        events = Events.fromFd(fd, bus.maxReplySize)
      end
    end
  end

  local self = DeviceBus.new("socket", rpc, payload, events)
  self.generation = hello.gen
  return self
end

local function sleep(milliseconds)
  time.nanosleep({ tv_sec = milliseconds // 1000,
                   tv_nsec = (milliseconds % 1000) * 1000 * 1000 })
end

function bus.connect(options)
  options = options or {}
  local roles = options.roles or defaultRoles

  local path = options.socketPath or os.getenv("OC2_BUS_SOCKET") or bus.socketPath
  local useSocket = path ~= "none" and path ~= ""
  local socketReason, directReason

  for attempt = 1, bus.connectAttempts do
    if useSocket then
      local connected, reason = connectSocket(path, roles)
      if connected then
        return connected
      end
      socketReason = reason
    end

    local direct
    direct, directReason = connectPorts()
    if direct then
      return direct
    end

    if attempt < bus.connectAttempts then
      sleep(bus.connectRetryDelay)
    end
  end

  if socketReason then
    return nil, string.format("could not reach the bus daemon (%s), nor the ports directly (%s)",
                              tostring(socketReason), tostring(directReason))
  end
  return nil, tostring(directReason)
end

bus.DeviceBus = DeviceBus
bus.Device = Device
bus.null = DeviceBus.null

return bus
